### PR TITLE
Add template support to KafkaUser resource to manage secret metadata

### LINKED ISSUE #3237: [Enhancement] KafkaUser - Pass annotations to secrets
When I create a KafkaUser I like to define annotations on it, which will be passed/copied to the secrets created by the operator.

This would enable use cases like https://github.com/EmberStack/kubernetes-reflector or custom implementations on secrets.

Is there a way to do this?

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Enhancement / new feature

### Description

This PR adds template support to the KafkaUser resource to manage metadata (labels & annos) of the user secret created by the UO. As part of this change, the `mergeLabelsOrAnnotations` method was moved from `AbstractModel` to `Utils` class in `operator-common` so that it can be reused in the UO. I also wrote new tests for this method.

This should close #3237.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@
 * Migration to Helm 3
 * Refactored the format of the `KafkaRebalance` resource's status. The state of the rebalance is now displayed in the associated `Condition`'s `type` field rather than the `status` field. This was done so that the information would display correctly in various Kubernetes tools.
 * Use Strimzi Kafka Bridge 0.17.0
+* Make it possible to configure labels and annotations for secrets created by User Operator
```
Comment (samuel-hawker) [inline CHANGELOG.md:20]: `created by the Kafka User operator`?

Comment (scholzj) [reply CHANGELOG.md:20]: It is called just 'User Operator' everywhere. So I think this is ok. Or was it the 'the' you wanted to add?

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaUserTemplate.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import io.strimzi.crdgenerator.annotations.DescriptionFile;
+import io.sundr.builder.annotations.Buildable;
+import lombok.EqualsAndHashCode;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Representation of a template for a KafkaUser resources.
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaUserTemplate.java:21]:
```suggestion
 * Representation of a template for a KafkaUser resource.
```

File: api/src/main/java/io/strimzi/api/kafka/model/template/KafkaUserTemplate.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+@DescriptionFile
+@EqualsAndHashCode
+public class KafkaUserTemplate implements Serializable, UnknownPropertyPreserving {
+    private static final long serialVersionUID = 1L;
+
+    private ResourceTemplate secret;
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
+
+    @Description("Template for KafkaUser resources. " +
+            "The template allows users to specify how is the `Secret` with password or certificates are generated.")
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/template/KafkaUserTemplate.java:38]:
```suggestion
            "The template allows users to specify how the `Secret` with password or TLS certificates is generated.")
```

File: CHANGELOG.md (hunk 1 above)
Comment (scholzj) [reply CHANGELOG.md]: I added the `the`.

File: documentation/api/io.strimzi.api.kafka.model.template.KafkaUserTemplate.adoc [hunk 4]
```diff
@@ -0,0 +1,23 @@
+`KafkaUserTemplate` can be used specify additional labels and annotation for the secret with user password or certificate created by the User Operator.
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.template.KafkaUserTemplate.adoc:1]:
```suggestion
`KafkaUserTemplate` is used specify additional labels and annotations for the secret created by the User Operator.
```

File: helm-charts/helm2/strimzi-kafka-operator/templates/044-Crd-kafkauser.yaml [hunk 5]
```diff
@@ -143,6 +143,24 @@ spec:
...
+              properties:
+                secret:
+                  type: object
+                  properties:
+                    metadata:
+                      type: object
+                      properties:
+                        labels:
+                          type: object
+                          description: Labels which should be added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```
Comment (PaulRMellor) [inline helm-charts/helm2/strimzi-kafka-operator/templates/044-Crd-kafkauser.yaml:157]:
```suggestion
                          description: Labels added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```

File: helm-charts/helm2/strimzi-kafka-operator/templates/044-Crd-kafkauser.yaml [hunk 6]
```diff
@@ -143,6 +143,24 @@ spec:
...
+                        annotations:
+                          type: object
+                          description: Annotations which should be added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```
Comment (PaulRMellor) [inline helm-charts/helm2/strimzi-kafka-operator/templates/044-Crd-kafkauser.yaml:160]:
```suggestion
                          description: Annotations added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```

File: helm-charts/helm2/strimzi-kafka-operator/templates/044-Crd-kafkauser.yaml [hunk 7]
```diff
@@ -143,6 +143,24 @@ spec:
...
+                      description: Metadata which should be applied to the resource.
```
Comment (PaulRMellor) [inline helm-charts/helm2/strimzi-kafka-operator/templates/044-Crd-kafkauser.yaml:161]:
```suggestion
                      description: Metadata applied to the resource.
```

File: helm-charts/helm3/strimzi-kafka-operator/crds/044-Crd-kafkauser.yaml [hunk 8]
```diff
@@ -139,6 +139,24 @@ spec:
...
+              properties:
+                secret:
+                  type: object
+                  properties:
+                    metadata:
+                      type: object
+                      properties:
+                        labels:
+                          type: object
+                          description: Labels which should be added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```
Comment (PaulRMellor) [inline helm-charts/helm3/strimzi-kafka-operator/crds/044-Crd-kafkauser.yaml:153]:
```suggestion
                          description: Labels added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```

File: helm-charts/helm3/strimzi-kafka-operator/crds/044-Crd-kafkauser.yaml [hunk 9]
```diff
@@ -139,6 +139,24 @@ spec:
...
+                        annotations:
+                          type: object
+                          description: Annotations which should be added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```
Comment (PaulRMellor) [inline helm-charts/helm3/strimzi-kafka-operator/crds/044-Crd-kafkauser.yaml:156]:
```suggestion
                          description: Annotations added to the resource template. Can be applied to different resources such as `StatefulSets`, `Deployments`, `Pods`, and `Services`.
```

File: helm-charts/helm3/strimzi-kafka-operator/crds/044-Crd-kafkauser.yaml [hunk 10]
```diff
@@ -139,6 +139,24 @@ spec:
...
+                      description: Metadata which should be applied to the resource.
```
Comment (PaulRMellor) [inline helm-charts/helm3/strimzi-kafka-operator/crds/044-Crd-kafkauser.yaml:157]:
```suggestion
                      description: Metadata applied to the resource.
```

File: install/cluster-operator/044-Crd-kafkauser.yaml [hunk 11]
```diff
@@ -163,6 +163,30 @@ spec:
...
+              properties:
+                secret:
+                  type: object
+                  properties:
+                    metadata:
+                      type: object
+                      properties:
+                        labels:
+                          type: object
+                          description: Labels which should be added to the resource
```
Comment (PaulRMellor) [inline install/cluster-operator/044-Crd-kafkauser.yaml:177]:
```suggestion
                          description: Labels added to the resource
```

File: install/cluster-operator/044-Crd-kafkauser.yaml [hunk 12]
```diff
@@ -163,6 +163,30 @@ spec:
...
+                            template. Can be applied to different resources such as
+                            `StatefulSets`, `Deployments`, `Pods`, and `Services`.
+                        annotations:
+                          type: object
+                          description: Annotations which should be added to the resource
```
Comment (PaulRMellor) [inline install/cluster-operator/044-Crd-kafkauser.yaml:182]:
```suggestion
                          description: Annotations added to the resource
```

File: install/cluster-operator/044-Crd-kafkauser.yaml [hunk 13]
```diff
@@ -163,6 +163,30 @@ spec:
...
+                            template. Can be applied to different resources such as
+                            `StatefulSets`, `Deployments`, `Pods`, and `Services`.
+                      description: Metadata which should be applied to the resource.
```
Comment (PaulRMellor) [inline install/cluster-operator/044-Crd-kafkauser.yaml:185]:
```suggestion
                      description: Metadata applied to the resource.
```

File: install/user-operator/04-Crd-kafkauser.yaml [hunk 14]
```diff
@@ -163,6 +163,30 @@ spec:
...
+              properties:
+                secret:
+                  type: object
+                  properties:
+                    metadata:
+                      type: object
+                      properties:
+                        labels:
+                          type: object
+                          description: Labels which should be added to the resource
```
Comment (PaulRMellor) [inline install/user-operator/04-Crd-kafkauser.yaml:177]:
```suggestion
                          description: Labels added to the resource
```

File: install/user-operator/04-Crd-kafkauser.yaml [hunk 15]
```diff
@@ -163,6 +163,30 @@ spec:
...
+                            template. Can be applied to different resources such as
+                            `StatefulSets`, `Deployments`, `Pods`, and `Services`.
+                        annotations:
+                          type: object
+                          description: Annotations which should be added to the resource
```
Comment (PaulRMellor) [inline install/user-operator/04-Crd-kafkauser.yaml:182]:
```suggestion
                          description: Annotations added to the resource
```

File: install/user-operator/04-Crd-kafkauser.yaml [hunk 16]
```diff
@@ -163,6 +163,30 @@ spec:
...
+                            template. Can be applied to different resources such as
+                            `StatefulSets`, `Deployments`, `Pods`, and `Services`.
+                      description: Metadata which should be applied to the resource.
```
Comment (PaulRMellor) [inline install/user-operator/04-Crd-kafkauser.yaml:185]:
```suggestion
                      description: Metadata applied to the resource.
```

**Comment by scholzj** [comment]
@PaulRMellor Most of your comments are related to the pre-existing MetadataTemplate class which I just use here. So they need to be fixed on many more places. I will merge this as is and open a new PR for the comments related to the other parts to make sure we keep clean PRs.

**Comment by scholzj** [comment]
@PaulRMellor I opened #3284

**Pull request merged**